masonry/src/tree_arena.rs

```diff
@@ -76,22 +63,28 @@ pub struct ArenaMut<'a, Item> {
 
 /// A reference type giving shared access to an item's children.
 ///
-/// When you borrow an item from a [`TreeArena`], you get two values, returned
-/// separately for lifetime reasons: a reference to the item itself, and a token
-/// to access its children.
+/// See [`ArenaRef`] for more information.
 pub struct ArenaRefChildren<'a, Item> {
     id: Option<u64>,
     children: &'a Vec<TreeNode<Item>>,
+    parents_map: ArenaMapRef<'a>,
 }
 
 /// A reference type giving mutable access to an item's children.
 ///
-/// When you borrow an item from a [`TreeArena`], you get two values, returned
-/// separately for lifetime reasons: a reference to the item itself, and a token
-/// to access its children.
+/// See [`ArenaMut`] for more information.
 pub struct ArenaMutChildren<'a, Item> {
     id: Option<u64>,
     children: &'a mut Vec<TreeNode<Item>>,
+    parents_map: ArenaMapMut<'a>,
+}
+
+#[derive(Clone, Copy)]
+pub struct ArenaMapRef<'a> {
```

Comment:
Would this be clearer if it was just:
```rust
pub struct ArenaMap(HashMap<u64, Option<u64>>);
```
then using `&` and `&mut` references as applicable.

Reply:
That's not a bad idea in principle, but the idea is that ArenaMapRef represents a capability to the "structure" of the arena. In future PRs it'll be passed as arguments, and in a future arena implementation, it'll be a stateless token.